### PR TITLE
Import pem

### PR DESCRIPTION
**Pull request by mattgle**
## Summary

- Add feature to import accounts using .pem as file or string.
- Added a few TODOs to use Toast Messages in some Error cases.

## Screenshots

https://user-images.githubusercontent.com/44204622/195902437-2123e666-d439-484f-a430-faf9ed44ac30.mp4

https://user-images.githubusercontent.com/44204622/195902536-925101a9-3c90-42a7-83cd-c9db533dff17.mp4

## Github Issue

https://github.com/Psychedelic/plug-mobile/issues/494

### CODE REVIEW

File: src/redux/slices/keyring.ts [hunk 1]
```diff
@@ -200,6 +200,58 @@ export const login = createAsyncThunk(
...
+      const instance = KeyRing.getInstance();
+      const response = await instance?.importAccountFromPem({
+        pem,
+        icon,
+        name,
+      });
+      return response;
+    } catch (e: any) {
+      // TODO: Add toast to handle error.
+      return rejectWithValue({ error: e.message });
```
Comment (ignaciosantise) [inline src/redux/slices/keyring.ts:219]: return the error directly, i've change that in the other actions
` return rejectWithValue(e.message);`

File: src/redux/slices/keyring.ts [hunk 2]
```diff
@@ -200,6 +200,58 @@ export const login = createAsyncThunk(
...
+  ) => {
+    try {
+      const instance = KeyRing.getInstance();
+      const response = await instance?.validatePem({ pem });
+      if (response) {
+        onSuccess();
+      } else {
+        onFailure();
+      }
+      onFinish();
```
Comment (ignaciosantise) [inline src/redux/slices/keyring.ts:248]: what's the point of the `onFinish`?

File: src/screens/tabs/Profile/screens/Accounts/index.js [hunk 3]
```diff
@@ -168,6 +170,10 @@ const Accounts = ({ modalRef, onClose, ...props }) => {
             accountsModalRef={modalRef}
             account={selectedAccount}
           />
+          <CreateImportAccount
+            modalRef={createImportAccountRef}
+            accountsModal={modalRef}
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/Accounts/index.js:175]: `accountsModalRef={modalRef}`

File: src/screens/tabs/Profile/screens/CreateEditAccount/index.js [hunk 4]
```diff
@@ -9,33 +10,56 @@ import UserIcon from '@/commonComponents/UserIcon';
...
+  accountsModalRef,
+  pem,
+  createImportModalRef,
+}) => {
   const { t } = useTranslation();
   const editEmojiRef = useRef(null);
   const [accountName, setAccountName] = useState('');
   const [editTouched, setEditTouched] = useState(false);
   const [emoji, setEmoji] = useState('');
+  const dispatch = useDispatch();
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/CreateEditAccount/index.js:37]: useAppDispatch

File: src/screens/tabs/Profile/screens/CreateEditAccount/index.js [hunk 5]
```diff
@@ -9,33 +10,56 @@ import UserIcon from '@/commonComponents/UserIcon';
...
+  editSubaccount,
+  importAccountFromPem,
+} from '@/redux/slices/keyring';
 
 import EditEmoji from '../EditEmoji';
 import styles from './styles';
 
-const CreateEditAccount = ({ modalRef, account, accountsModalRef }) => {
+/**
+ * @param {{ modalRef: any, accountsModalRef?: any, account?: any, pem?: string, createImportModalRef?: any } param
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/CreateEditAccount/index.js:23]: define the real interfaces here, or we could move the file to TS, maybe its not that hard

File: src/redux/slices/keyring.ts (hunk 2 above)
Comment (mattgle) [reply src/redux/slices/keyring.ts:248]: Im using it to set a loding in false

File: src/screens/tabs/Profile/screens/CreateEditAccount/index.js [hunk 6]
```diff
@@ -9,33 +10,56 @@ import UserIcon from '@/commonComponents/UserIcon';
...
+            walletId: account.walletId,
+            ...nameAndIcon,
+          })
+        : pem
+        ? importAccountFromPem({
+            ...nameAndIcon,
+            pem,
+          })
+        : createSubaccount(nameAndIcon)
+    );
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/CreateEditAccount/index.js:57]: i think it'll be more readable if we replace the double ternary with ifs

if(account) -> dispatch
else if(pem) -> dispatch
else -> dispatch

File: src/screens/tabs/Profile/screens/CreateEditAccount/index.js (hunk 5 above)
Comment (mattgle) [reply src/screens/tabs/Profile/screens/CreateEditAccount/index.js:23]: I was on it!

File: src/screens/tabs/Profile/screens/CreateImportAccount/index.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    openImportKeyModal,
+  });
+
+  return (
+    <Modal adjustToContentHeight modalRef={modalRef}>
+      <Header
+        right={
+          <Text style={styles.headerAction} onPress={closeModal}>
+            {t('common.close')}
+          </Text>
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/CreateImportAccount/index.tsx:86]: there's a custom component used for this and the back button

File: src/screens/tabs/Profile/screens/ImportKey/index.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    );
+  };
+
+  return (
+    <Modal adjustToContentHeight modalRef={modalRef}>
+      <Header
+        right={
+          <Text style={styles.headerAction} onPress={closeModal}>
+            {t('common.close')}
+          </Text>
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/ImportKey/index.tsx:62]: same here

File: src/screens/tabs/Profile/screens/ImportKey/index.tsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  createImportRef: RefObject<Modalize>;
+  accountsModalRef: RefObject<Modalize>;
+}
+
+function ImportKey({ createImportRef, modalRef, accountsModalRef }: Props) {
+  const createEditAccount = useRef<Modalize>(null);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(false);
+  const [key, setKey] = useState('');
+  const dispatch = useDispatch();
```
Comment (ignaciosantise) [inline src/screens/tabs/Profile/screens/ImportKey/index.tsx:25]: useAppDispatch

File: src/constants/theme.js [hunk 10]
```diff
@@ -108,6 +108,7 @@ export const Shadow = StyleSheet.create({
 // TODO: Remove styles that doesnt match with the style guide in figma, and remove colors
 export const FontStyles = StyleSheet.create({
   Base: fontMaker({ size: 14, weight: REGULAR }),
+  Error: fontMaker({ size: 14, weight: MEDIUM, color: Colors.Red }),
```
Comment (ignaciosantise) [inline src/constants/theme.js:111]: this style is not defined in the style guide, should we add it here? 🤔

**Review by ignaciosantise — APPROVED** [review]
🚀

**Pull request merged**